controller/query_blackbox_test.go

```diff
@@ -0,0 +1,347 @@
+			compareWithGoldenUUIDAgnostic(t, filepath.Join(rest.testDir, "create", "ok_create.req.payload.golden.json"), cq)
+		})
+	})
+
+	rest.T().Run("fail", func(t *testing.T) {
+		t.Run("Unauthorized", func(t *testing.T) {
+			fxt := tf.NewTestFixture(t, rest.DB,
+				tf.CreateWorkItemEnvironment())
+			cq := getQueryCreatePayload("query 1", nil)
+			svc, ctrl := rest.UnSecuredController()
+			// when
+			test.CreateQueryUnauthorized(t, svc.Context, svc, ctrl, fxt.Spaces[0].ID, cq)
+		})
+		t.Run("invalid query", func(t *testing.T) {
+			fxt := tf.NewTestFixture(t, rest.DB,
+				tf.CreateWorkItemEnvironment())
+			cq := getQueryCreatePayload("query 1", nil)
+			cq.Data.Attributes.Fields = `{"invalid: json"}`
+			svc, ctrl := rest.SecuredControllerWithIdentity(fxt.Identities[0])
+			// when
+			test.CreateQueryBadRequest(t, svc.Context, svc, ctrl, fxt.Spaces[0].ID, cq)
+		})
+		t.Run("conflict same title, spaceID, creator", func(t *testing.T) {
+			fxt := tf.NewTestFixture(t, rest.DB,
+				tf.CreateWorkItemEnvironment(),
+				tf.Queries(1, tf.SetQueryTitles("q1")))
+			cq := getQueryCreatePayload(fxt.Queries[0].Title, nil)
+			svc, ctrl := rest.SecuredControllerWithIdentity(fxt.Identities[0])
+			// when
+			test.CreateQueryConflict(t, svc.Context, svc, ctrl, fxt.Spaces[0].ID, cq)
+		})
+		t.Run("unknown space ID", func(t *testing.T) {
+			// given
+			fxt := tf.NewTestFixture(t, rest.DB, tf.CreateWorkItemEnvironment())
+			cq := getQueryCreatePayload("new query", nil)
+			svc, ctrl := rest.SecuredControllerWithIdentity(fxt.Identities[0])
+			// when
+			test.CreateQueryNotFound(t, svc.Context, svc, ctrl, uuid.NewV4(), cq)
+		})
+	})
+}
+
+func (rest *TestQueryREST) TestList() {
+	resetFn := rest.DisableGormCallbacks()
+	defer resetFn()
+
+	rest.T().Run("success", func(t *testing.T) {
+		t.Run("ok", func(t *testing.T) {
+			fxt := tf.NewTestFixture(t, rest.DB,
+				tf.CreateWorkItemEnvironment(),
+				tf.Queries(3, tf.SetQueryTitles("q1", "q2", "q3")))
+			fxt2 := tf.NewTestFixture(t, rest.DB,
+				tf.CreateWorkItemEnvironment(),
+				tf.Queries(3, tf.SetQueryTitles("q4", "q5", "q6")))
+			svc, ctrl := rest.SecuredControllerWithIdentity(fxt.Identities[0])
+			// when
+			_, qList := test.ListQueryOK(t, svc.Context, svc, ctrl, fxt.Spaces[0].ID, nil, nil)
+			// then
+			require.NotNil(t, qList)
+			mustHave := map[string]struct{}{
+				"q1": {},
+				"q2": {},
+				"q3": {},
+			}
+			for _, q := range qList.Data {
+				delete(mustHave, q.Attributes.Title)
+			}
+			assert.Empty(t, mustHave)
+			assert.Equal(t, 3, qList.Meta.TotalCount)
+			// list by different user
+			// when
+			svc, ctrl = rest.SecuredControllerWithIdentity(fxt2.Identities[0])
+			_, qList = test.ListQueryOK(t, svc.Context, svc, ctrl, fxt2.Spaces[0].ID, nil, nil)
+			// then
+			require.NotNil(t, qList)
+			mustHave = map[string]struct{}{
+				"q4": {},
+				"q5": {},
+				"q6": {},
+			}
+			for _, q := range qList.Data {
+				delete(mustHave, q.Attributes.Title)
```

Comment:
Hey, I just [realized](https://golang.org/pkg/builtin/#delete) that you can call `delete` before checking if the key exists. Great.